plugins/extractors/bigtable/middleware.go

```diff
@@ -0,0 +1,107 @@
+package bigtable
+
+import (
+	"context"
+	"time"
+
+	"cloud.google.com/go/bigtable"
+	"github.com/goto/meteor/utils"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+)
+
+type AdminClientMW struct {
+	tableDuration  metric.Int64Histogram
+	tablesDuration metric.Int64Histogram
+	projectID      string
+	instanceName   string
+	next           AdminClient
+}
+
+type InstancesAdminClientMW struct {
+	instancesDuration metric.Int64Histogram
+	projectID         string
+	next              InstanceAdminClient
+}
+
+func WithAdminClientMW(next AdminClient, projectID, instanceName string) (AdminClient, error) {
+	meter := otel.Meter("")
+
+	tablesDuration, err := meter.Int64Histogram("meteor.bigtable.client.tables.duration", metric.WithUnit("ms"))
+	if err != nil {
+		return nil, err
+	}
+
+	tableDuration, err := meter.Int64Histogram("meteor.bigtable.client.table.duration", metric.WithUnit("ms"))
+	if err != nil {
+		return nil, err
+	}
+
+	return &AdminClientMW{
+		tableDuration:  tableDuration,
+		tablesDuration: tablesDuration,
+		projectID:      projectID,
+		instanceName:   instanceName,
+		next:           next,
+	}, nil
+}
+
+func WithInstancesAdminClientMW(next InstanceAdminClient, projectID string) (InstanceAdminClient, error) {
+	meter := otel.Meter("")
+
+	instancesDuration, err := meter.Int64Histogram("meteor.bigtable.client.instances.duration", metric.WithUnit("ms"))
+	if err != nil {
+		return nil, err
+	}
+
+	return &InstancesAdminClientMW{
+		instancesDuration: instancesDuration,
+		projectID:         projectID,
+		next:              next,
+	}, nil
+}
+
+func (o *AdminClientMW) Tables(ctx context.Context) (res []string, err error) {
+	defer func(start time.Time) {
+		o.tablesDuration.Record(ctx,
+			time.Since(start).Milliseconds(),
+			metric.WithAttributes(
+				attribute.String("bt.project_id", o.projectID),
+				attribute.String("bt.error_code", utils.StatusCode(err).String()),
```

Comment:
It might be better to rely on the [`*APIError.Reason()`](https://pkg.go.dev/github.com/googleapis/gax-go/v2/apierror#APIError.Reason). The documentation for the field [`errdetails.ErrorInfo.Reason`](https://github.com/googleapis/go-genproto/blob/23370e0/googleapis/rpc/errdetails/error_details.pb.go#L68-L73):

> ```go
>	// The reason of the error. This is a constant value that identifies the
>	// proximate cause of the error. Error reasons are unique within a particular
>	// domain of errors. This should be at most 63 characters and match a
>	// regular expression of `[A-Z][A-Z0-9_]+[A-Z0-9]`, which represents
>	// UPPER_SNAKE_CASE.
>	Reason string `protobuf:"bytes,1,opt,name=reason,proto3" json:"reason,omitempty"`
> ```

This reason is parsed from the gRPC status details that could have information specific to Bigtable rather than generic gRPC codes - https://github.com/googleapis/gax-go/blob/main/v2/apierror/apierror.go#L255

Reply:
updated